src/util/psbt/mod.rs

```diff
@@ -209,6 +215,366 @@ impl PartiallySignedTransaction {
+        // We check this everywhere but fail early if `input_index` is invalid.
+        self.check_index_is_within_bounds(input_index)?;
+
+        let algo = self.signing_algorithm(input_index)?;
+        match (algo, key) {
+            (SigningAlgorithm::Ecdsa, PrivateKeyOrPair::Key(k)) => self.sign_ecdsa(k, input_index, cache, secp),
+            (SigningAlgorithm::Schnorr, PrivateKeyOrPair::Pair(_)) => unimplemented!("taproot PSBT signing not yet implemented"),
+            (_, _) => Err(SighashError::MismatchedAlgoKey),
+        }
+    }
+
+    fn sign_ecdsa<C: Signing, T: Deref<Target=Transaction>>(
+        &mut self,
+        sk: PrivateKey,
+        input_index: usize,
+        cache: Option<&mut SighashCache<T>>,
+        secp: &Secp256k1<C>
+    ) -> Result<(), SighashError> {
+        let (msg, sighash_ty) = self.sighash_ecdsa(input_index, cache)?;
+
+        let sig = EcdsaSig {
+            sig: secp.sign_ecdsa(&msg, &sk.inner),
+            hash_ty: sighash_ty,
+        };
+
+        let pk = sk.public_key(secp);
+        self.inputs[input_index]
+            .partial_sigs
+            .insert(pk, sig);
+
+        Ok(())
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    fn sighash_ecdsa<T: Deref<Target=Transaction>>(
+        &self,
+        input_index: usize,
+        cache: Option<&mut SighashCache<T>>,
+    ) -> Result<(Message, EcdsaSighashType), SighashError> {
+        match cache {
+            Some(cache) => self._sighash_ecdsa(input_index, cache),
+            None => {
+                let mut cache = SighashCache::new(&self.unsigned_tx);
+                self._sighash_ecdsa(input_index, &mut cache)
+            }
+        }
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    fn _sighash_ecdsa<T: Deref<Target=Transaction>>(
+        &self,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+    ) -> Result<(Message, EcdsaSighashType), SighashError> {
+        assert_eq!(self.signing_algorithm(input_index), Ok(SigningAlgorithm::Ecdsa));
+
+        let input = self.checked_input(input_index)?;
+        let utxo = self.spend_utxo(input_index)?;
+        let script = &utxo.script_pubkey; // scriptPubkey for input spend utxo.
+
+        let hash_ty = input
+            .sighash_type
+            .map(|ty| ty.ecdsa_hash_ty())
+            .unwrap_or(Ok(EcdsaSighashType::All))
+            .map_err(|_| SighashError::InvalidSighashType)?; // Only support standard sighash types.
+
+        let output_type = self.output_type(input_index)?;
+
+        let is_wpkh = output_type == OutputType::Wpkh;
+        let is_nested_wpkh = output_type == OutputType::ShWpkh;
+
+        let sighash = if output_type.is_segwit() {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
+                    Script::p2wpkh_script_code(&script).ok_or(SighashError::NotWpkh)?
```

Comment:
(here and below)Part of me wants to unwrap here. But I also understand it's probably much easier to reason about this than the unwrapping

Reply:
I do not understand the error path fully enough to judge, the error does not seem to make the code _less_ readable or debugable, so I'm leaning towards leaving it in.